core/unit_test/tools/TestLogicalSpaces.hpp

```diff
@@ -131,7 +131,7 @@ void test_malloc_free() {
   auto* temp =
       Kokkos::kokkos_malloc<fake_memory_space>("does_malloc_work", 1000);
   expect_deallocation_event("does_malloc_work", "TestSpace", "Error in free");
-  Kokkos::kokkos_free(temp);
+  Kokkos::kokkos_free<fake_memory_space>(temp);
```

Comment:
https://github.com/kokkos/kokkos/blob/9764d0377982bdd31c9481cd6067de40ff8eb6dc/core/src/Kokkos_Core.hpp#L173-L174
deduces the wrong space and leads to the illegal downcast